### ISSUE TITLE
"Bad parameter specification" while reading some models

### ISSUE THREAD

**Issue by anhinga** [clab/dynet#1430]
I believe that this happens when a model has some FULL_GRAD lines after some ZERO_GRAD lines.

I think the reason is in function "read_param_header", lines 41-53.

https://github.com/clab/dynet/blob/master/dynet/io.cc

called from line 157 of the same file. The logic here is such that "zero_grad" is initialized to "false" on line 147 (outside of the while loop), but once it is set to "true" by "read_param_header", there is no way for it to become "false" again, hence the bug (it breaks on trying to interpret a line of gradient values as a parameter header).

**Comment by anhinga** [comment]
I think the correct fix is likely to consist of moving the code `bool zero_grad = false;` in line 147 of

https://github.com/clab/dynet/blob/master/dynet/io.cc

to the line following `while(std::getline(datastream, line)) {` (currently line 156).

Meanwhile, if one does not want to rebuild DyNet, the workaround is to remove gradient information from the model file. The following script works for me in Python 3.6 for the latest format of DyNet models:

```python
f = open("YOUR MODEL FILE NAME")

while True:
  line = f.readline()
  if line == "":
    break
  parameters = line.split()
  if parameters[-1] == "FULL_GRAD":
    n = int(parameters[-2])
    parameters[-2] = str(n//2)
    parameters[-1] = "ZERO_GRAD"
    line = " ".join(parameters)
    print(line)
    line = f.readline()
    print(line, end="")
    line = f.readline() # skipping gradient line
  else:
    print(line, end="")
    line = f.readline()
    print(line, end="")
```

**Comment by pmichel31415** [comment]
Hi @anhinga sorry for the late reply and thanks for spotting this and providing a workaround. Do you think you could submit a pull request incorporating your changes and (if you have time) a relevant test case in `tests/python/test.py`?

That would be extremely helpful!

**Comment by anhinga** [comment]
Hi @pmichel31415 

I have to admit that I did not master the art of building DyNet yet :-) I am just using the latest version from "pip install". 

Speaking about the workaround script, the extent to which it was tested is that I had an 800 megabyte model, and after applying the workaround I was able to load it, and the test run gave the same performance numbers as the numbers obtained on the fly during the training run which generated that model. But I don't have any small regression test for it.

The suggestion to fix the underlying problem by moving line 147
```python
bool zero_grad = false;
```

between the lines 156-157
```python
while(std::getline(datastream, line)) {
read_param_header(line, type, name, dim, byte_count, zero_grad);
```

is probably incomplete, as the same code pattern also exists in different versions of `TextFileLoader::populate` function further down in

https://github.com/clab/dynet/blob/master/dynet/io.cc

The existing logic comes from your Jul 5, 2017 commit "Initial commit', so you are probably the best person to evaluate whether this particular suggestion makes sense...

**Comment by pmichel31415** [comment]
I understand, thanks for the details, I should be able to fix this. One more thing, do you have a minimal reproducible piece of code that I could use as a test? In python or C++, whatever you're more familiar with

**Comment by anhinga** [comment]
For the Python script, the context is quite big. I took this wonderful project

https://github.com/jcyk/greedyCWS

upgraded it to DyNet 2 (because I had no idea how to build DyNet 1.1, so it was easier for me to upgrade the greedyCWS), and then I've run it on some training corpus I had, and when I wanted to load the resulting model (which was about 800 megabytes), I've run into that issue.

(The way to reproduce the situation in the small is to create a model, when some lines with FULL_GRAD marker follow some lines with ZERO_GRAD marker, that would trigger the error. The workaround script just cleans all the gradient info from the model.)

**Comment by pmichel31415** [comment]
Thank you so much for all the details. This will be fixed in #1450

**Comment by anhinga** [comment]
Thanks!
